File: filter-contacts.js
import https from 'node:https';
import { loadEnv, sleep, readJSON, writeJSON } from './lib.js';

const env = loadEnv();
const HUBSPOT_TOKEN = env.HUBSPOT_API_TOKEN;

const CONTACTS_FILE = 'data/contacts.json';
const FILTERED_FILE = 'data/filtered.json';
const EXCLUDED_FILE = 'data/excluded.json';
const PROGRESS_FILE = 'data/.filter-progress';

// --- Headline blocklist ---

const HEADLINE_PATTERNS = [
  // Consultants
  { pattern: /consultant/i, keyword: 'consultant' },
  { pattern: /consulting/i, keyword: 'consulting' },
  { pattern: /advisory/i, keyword: 'advisory' },
  { pattern: /advisor/i, keyword: 'advisor' },
  { pattern: /adviser/i, keyword: 'adviser' },
  // Freelancers
  { pattern: /freelance/i, keyword: 'freelance' },
  { pattern: /freelancer/i, keyword: 'freelancer' },
  { pattern: /independent/i, keyword: 'independent' },
  // Agency
  { pattern: /agency/i, keyword: 'agency' },
  { pattern: /we help/i, keyword: 'we help' },
  { pattern: /we build/i, keyword: 'we build' },
  { pattern: /we engineer/i, keyword: 'we engineer' },
  { pattern: /we design/i, keyword: 'we design' },
  { pattern: /we create/i, keyword: 'we create' },
  { pattern: /we deliver/i, keyword: 'we deliver' },
  { pattern: /we scale/i, keyword: 'we scale' },
  { pattern: /helping\s+.*?\s*(companies|businesses|brands|startups|teams|agencies)/i, keyword: 'helping companies/businesses' },
  // Pitch-style / service seller language
  { pattern: /for hire/i, keyword: 'for hire' },
  { pattern: /available for/i, keyword: 'available for' },
  { pattern: /open to work/i, keyword: 'open to work' },
  { pattern: /seeking opportunities/i, keyword: 'seeking opportunities' },
  { pattern: /seeking for opportunities/i, keyword: 'seeking for opportunities' },
  { pattern: /grow(ing)? your/i, keyword: 'grow(ing) your' },
  { pattern: /scale your/i, keyword: 'scale your' },
  { pattern: /boost your/i, keyword: 'boost your' },
  { pattern: /your sales/i, keyword: 'your sales' },
  { pattern: /your business/i, keyword: 'your business' },
  { pattern: /your revenue/i, keyword: 'your revenue' },
  { pattern: /your growth/i, keyword: 'your growth' },
  { pattern: /your customers/i, keyword: 'your customers' },
  { pattern: /done for you/i, keyword: 'done for you' },
  { pattern: /done with you/i, keyword: 'done with you' },
  { pattern: /on autopilot/i, keyword: 'on autopilot' },
  { pattern: /book meetings/i, keyword: 'book meetings' },
  { pattern: /book demos/i, keyword: 'book demos' },
  { pattern: /lead gen\b/i, keyword: 'lead gen' },
  { pattern: /lead generation/i, keyword: 'lead generation' },
  { pattern: /your onboarding/i, keyword: 'your onboarding' },
  { pattern: /into revenue/i, keyword: 'into revenue' },
  { pattern: /into conversions/i, keyword: 'into conversions' },
  { pattern: /let's talk/i, keyword: "let's talk" },
  { pattern: /scaling\s+.*?(companies|businesses|startups|saas)/i, keyword: 'scaling companies' },
  { pattern: /\bcoach\b/i, keyword: 'coach' },
  { pattern: /\bspeaker\b/i, keyword: 'speaker' },
  { pattern: /\bevangelist\b/i, keyword: 'evangelist' },
  { pattern: /\bentrepreneur\b/i, keyword: 'entrepreneur' },
  { pattern: /\bempowering\b/i, keyword: 'empowering' },
  { pattern: /managing partner/i, keyword: 'managing partner' },
  { pattern: /revenue architects?/i, keyword: 'revenue architect(s)' },
  { pattern: /for visionary/i, keyword: 'for visionary' },
  { pattern: /solutions partner/i, keyword: 'solutions partner' },
  { pattern: /partner für/i, keyword: 'partner für' },
  { pattern: /try\s+\S+\.(com|io|co|ai)/i, keyword: 'try [product URL]' },
  // HubSpot ecosystem
  { pattern: /hubspot consultant/i, keyword: 'hubspot consultant' },
  { pattern: /hubspot partner/i, keyword: 'hubspot partner' },
  { pattern: /hubspot certified/i, keyword: 'hubspot certified' },
  { pattern: /hubspot expert/i, keyword: 'hubspot expert' },
  { pattern: /hubspot implementation/i, keyword: 'hubspot implementation' },
  { pattern: /hubspot app/i, keyword: 'hubspot app' },
  { pattern: /build(ing)? hubspot/i, keyword: 'build(ing) hubspot' },
  { pattern: /revops agency/i, keyword: 'revops agency' },
  { pattern: /revops fixer/i, keyword: 'revops fixer' },
  { pattern: /martech agency/i, keyword: 'martech agency' },
  { pattern: /crm consultant/i, keyword: 'crm consultant' },
  { pattern: /crm implementation/i, keyword: 'crm implementation' },
];

function checkHeadline(headline) {
  if (!headline) return null;
  for (const { pattern, keyword } of HEADLINE_PATTERNS) {
    if (pattern.test(headline)) return keyword;
  }
  return null;
}

// --- HubSpot search ---

function searchHubSpot(query) {
  return new Promise((resolve, reject) => {
    const payload = JSON.stringify({
      query,
      limit: 10,
      properties: ['firstname', 'lastname'],
    });
    const opts = {
      hostname: 'api.hubapi.com',
      path: '/crm/v3/objects/contacts/search',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${HUBSPOT_TOKEN}`,
      },
    };
    const req = https.request(opts, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        if (res.statusCode === 429) {
          reject(new Error('RATE_LIMITED'));
          return;
        }
        if (res.statusCode >= 400) {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          return;
        }
        try {
          resolve(JSON.parse(data));
        } catch {
          resolve({ total: 0, results: [] });
        }
      });
    });
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

async function checkHubSpotExists(firstName, lastName) {
  const query = `${firstName} ${lastName}`;
  const data = await searchHubSpot(query);

  if (!data.total || data.total === 0) return null;

  for (const result of data.results || []) {
    const props = result.properties || {};
    const hsFirst = (props.firstname || '').toLowerCase();
    const hsLast = (props.lastname || '').toLowerCase();
    if (hsFirst === firstName.toLowerCase() && hsLast === lastName.toLowerCase()) {
      return result.id;
    }
  }
  return null;
}

// --- Main ---

async function main() {
  const contacts = readJSON(CONTACTS_FILE);
  if (contacts.length === 0) {
    console.error('No contacts found. Run scrape-contacts.js first.');
    process.exit(1);
  }

  let filtered = readJSON(FILTERED_FILE);
  let excluded = readJSON(EXCLUDED_FILE);

  // Resume support: track which contacts have been processed
  const processedUrls = new Set([
    ...filtered.map((c) => c.public_profile_url),
    ...excluded.map((c) => c.public_profile_url),
  ]);

  const remaining = contacts.filter((c) => !processedUrls.has(c.public_profile_url));
  console.log(`Total: ${contacts.length} | Already processed: ${processedUrls.size} | Remaining: ${remaining.length}`);

  let headlineExcluded = 0;
  let hubspotExcluded = 0;
  let passed = 0;

  for (let i = 0; i < remaining.length; i++) {
    const contact = remaining[i];
    const label = `[${processedUrls.size + i + 1}/${contacts.length}] ${contact.first_name} ${contact.last_name}`;

    // Filter A: Headline check
    const matchedKeyword = checkHeadline(contact.headline);
    if (matchedKeyword) {
      excluded.push({
        ...contact,
        reason: 'headline_filter',
        matched_keyword: matchedKeyword,
      });
      headlineExcluded++;
      console.log(`${label} — EXCLUDED (headline: "${matchedKeyword}")`);
      writeJSON(EXCLUDED_FILE, excluded);
      continue;
    }

    // Filter B: HubSpot dedup
    try {
      const hubspotId = await checkHubSpotExists(contact.first_name, contact.last_name);
      if (hubspotId) {
        excluded.push({
          ...contact,
          reason: 'already_in_hubspot',
          hubspot_id: hubspotId,
        });
        hubspotExcluded++;
        console.log(`${label} — EXCLUDED (already in HubSpot, ID: ${hubspotId})`);
        writeJSON(EXCLUDED_FILE, excluded);
        await sleep(100);
        continue;
      }
    } catch (err) {
      if (err.message === 'RATE_LIMITED') {
        console.error(`\nHubSpot RATE LIMITED. Halting. Re-run to resume.`);
        console.log(`Progress: ${headlineExcluded} headline + ${hubspotExcluded} hubspot excluded, ${passed} passed`);
        process.exit(1);
      }
      console.log(`${label} — HubSpot lookup error: ${err.message} (keeping contact)`);
    }

    // Passed both filters
    filtered.push(contact);
    passed++;
    console.log(`${label} — PASSED`);

    writeJSON(FILTERED_FILE, filtered);
    await sleep(100);
  }

  console.log(`\nDone.`);
  console.log(`  Total:                ${contacts.length}`);
  console.log(`  Excluded (headline):  ${headlineExcluded}`);
  console.log(`  Excluded (HubSpot):   ${hubspotExcluded}`);
  console.log(`  Ready for enrichment: ${passed}`);
  console.log(`  Total filtered:       ${filtered.length}`);
  console.log(`  Total excluded:       ${excluded.length}`);
}

main();

File: filter_contacts.py
#!/usr/bin/env python3
"""Filter scraped contacts by headline patterns and HubSpot dedup."""

import os
import re
import sys

import requests

from lib import init_env, read_json, sleep_ms, write_json

init_env()

HUBSPOT_TOKEN = os.environ.get('HUBSPOT_API_TOKEN', '')

CONTACTS_FILE = 'data/contacts.json'
FILTERED_FILE = 'data/filtered.json'
EXCLUDED_FILE = 'data/excluded.json'

# --- Headline blocklist ---

HEADLINE_PATTERNS = [
    # Consultants
    (re.compile(r'consultant', re.I), 'consultant'),
    (re.compile(r'consulting', re.I), 'consulting'),
    (re.compile(r'advisory', re.I), 'advisory'),
    (re.compile(r'advisor', re.I), 'advisor'),
    (re.compile(r'adviser', re.I), 'adviser'),
    # Freelancers
    (re.compile(r'freelance', re.I), 'freelance'),
    (re.compile(r'freelancer', re.I), 'freelancer'),
    (re.compile(r'independent', re.I), 'independent'),
    # Agency
    (re.compile(r'agency', re.I), 'agency'),
    (re.compile(r'we help', re.I), 'we help'),
    (re.compile(r'we build', re.I), 'we build'),
    (re.compile(r'we engineer', re.I), 'we engineer'),
    (re.compile(r'we design', re.I), 'we design'),
    (re.compile(r'we create', re.I), 'we create'),
    (re.compile(r'we deliver', re.I), 'we deliver'),
    (re.compile(r'we scale', re.I), 'we scale'),
    (re.compile(r'helping\s+.*?\s*(companies|businesses|brands|startups|teams|agencies)', re.I), 'helping companies/businesses'),
    # Pitch-style / service seller language
    (re.compile(r'for hire', re.I), 'for hire'),
    (re.compile(r'available for', re.I), 'available for'),
    (re.compile(r'open to work', re.I), 'open to work'),
    (re.compile(r'seeking opportunities', re.I), 'seeking opportunities'),
    (re.compile(r'seeking for opportunities', re.I), 'seeking for opportunities'),
    (re.compile(r'open for projects', re.I), 'open for projects'),
    (re.compile(r'open for clients', re.I), 'open for clients'),
    (re.compile(r'\bstrategist\b', re.I), 'strategist'),
    (re.compile(r'grow(ing)? your', re.I), 'grow(ing) your'),
    (re.compile(r'scale your', re.I), 'scale your'),
    (re.compile(r'boost your', re.I), 'boost your'),
    (re.compile(r'your sales', re.I), 'your sales'),
    (re.compile(r'your business', re.I), 'your business'),
    (re.compile(r'your revenue', re.I), 'your revenue'),
    (re.compile(r'your growth', re.I), 'your growth'),
    (re.compile(r'your customers', re.I), 'your customers'),
    (re.compile(r'done for you', re.I), 'done for you'),
    (re.compile(r'done with you', re.I), 'done with you'),
    (re.compile(r'on autopilot', re.I), 'on autopilot'),
    (re.compile(r'book meetings', re.I), 'book meetings'),
    (re.compile(r'book demos', re.I), 'book demos'),
    (re.compile(r'lead gen\b', re.I), 'lead gen'),
    (re.compile(r'lead generation', re.I), 'lead generation'),
    (re.compile(r'your onboarding', re.I), 'your onboarding'),
    (re.compile(r'into revenue', re.I), 'into revenue'),
    (re.compile(r'into conversions', re.I), 'into conversions'),
    (re.compile(r"let's talk", re.I), "let's talk"),
    (re.compile(r'\bi help\b', re.I), 'i help'),
    (re.compile(r'\binbound\b', re.I), 'inbound'),
    (re.compile(r'scaling\s+.*?(companies|businesses|startups|saas)', re.I), 'scaling companies'),
    (re.compile(r'\bcoach\b', re.I), 'coach'),
    (re.compile(r'\bspeaker\b', re.I), 'speaker'),
    (re.compile(r'\bevangelist\b', re.I), 'evangelist'),
    (re.compile(r'\bentrepreneur\b', re.I), 'entrepreneur'),
    (re.compile(r'\bempowering\b', re.I), 'empowering'),
    (re.compile(r'managing partner', re.I), 'managing partner'),
    (re.compile(r'revenue architects?', re.I), 'revenue architect(s)'),
    (re.compile(r'for visionary', re.I), 'for visionary'),
    (re.compile(r'solutions partner', re.I), 'solutions partner'),
    (re.compile(r'partner für', re.I), 'partner für'),
    (re.compile(r'try\s+\S+\.(com|io|co|ai)', re.I), 'try [product URL]'),
    # HubSpot ecosystem
    (re.compile(r'hubspot consultant', re.I), 'hubspot consultant'),
    (re.compile(r'hubspot\s+.*?partner', re.I), 'hubspot partner'),
    (re.compile(r'hubspot coach', re.I), 'hubspot coach'),
    (re.compile(r'hubspot certified', re.I), 'hubspot certified'),
    (re.compile(r'hubspot expert', re.I), 'hubspot expert'),
    (re.compile(r'hubspot implementation', re.I), 'hubspot implementation'),
    (re.compile(r'hubspot app', re.I), 'hubspot app'),
    (re.compile(r'build(ing)? hubspot', re.I), 'build(ing) hubspot'),
    (re.compile(r'revops agency', re.I), 'revops agency'),
    (re.compile(r'revops fixer', re.I), 'revops fixer'),
    (re.compile(r'martech agency', re.I), 'martech agency'),
    (re.compile(r'crm consultant', re.I), 'crm consultant'),
    (re.compile(r'crm implementation', re.I), 'crm implementation'),
]


def check_headline(headline):
    if not headline:
        return None
    for pattern, keyword in HEADLINE_PATTERNS:
        if pattern.search(headline):
            return keyword
    return None


# --- HubSpot search ---

def search_hubspot(query):
    resp = requests.post(
        'https://api.hubapi.com/crm/v3/objects/contacts/search',
        json={'query': query, 'limit': 10, 'properties': ['firstname', 'lastname']},
        headers={
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {HUBSPOT_TOKEN}',
        },
        timeout=30,
    )
    if resp.status_code == 429:
        raise Exception('RATE_LIMITED')
    resp.raise_for_status()
    return resp.json()


def check_hubspot_exists(first_name, last_name):
    query = f'{first_name} {last_name}'
    data = search_hubspot(query)

    if not data.get('total'):
        return None

    for result in data.get('results', []):
        props = result.get('properties', {})
        hs_first = (props.get('firstname') or '').lower()
        hs_last = (props.get('lastname') or '').lower()
        if hs_first == first_name.lower() and hs_last == last_name.lower():
            return result['id']
    return None


# --- Main ---

def main():
    contacts = read_json(CONTACTS_FILE)
    if not contacts:
        print('No contacts found. Run scrape_contacts.py first.')
        sys.exit(1)

    filtered = read_json(FILTERED_FILE)
    excluded = read_json(EXCLUDED_FILE)

    processed_urls = set(
        [c['public_profile_url'] for c in filtered] +
        [c['public_profile_url'] for c in excluded]
    )

    remaining = [c for c in contacts if c['public_profile_url'] not in processed_urls]
    print(f'Total: {len(contacts)} | Already processed: {len(processed_urls)} | Remaining: {len(remaining)}')

    headline_excluded = 0
    hubspot_excluded = 0
    passed = 0

    for i, contact in enumerate(remaining):
        label = f'[{len(processed_urls) + i + 1}/{len(contacts)}] {contact["first_name"]} {contact["last_name"]}'

        # Filter A: Headline check
        matched_keyword = check_headline(contact['headline'])
        if matched_keyword:
            excluded.append({
                **contact,
                'reason': 'headline_filter',
                'matched_keyword': matched_keyword,
            })
            headline_excluded += 1
            print(f'{label} — EXCLUDED (headline: "{matched_keyword}")')
            write_json(EXCLUDED_FILE, excluded)
            continue

        # Filter B: HubSpot dedup
        try:
            hubspot_id = check_hubspot_exists(contact['first_name'], contact['last_name'])
            if hubspot_id:
                excluded.append({
                    **contact,
                    'reason': 'already_in_hubspot',
                    'hubspot_id': hubspot_id,
                })
                hubspot_excluded += 1
                print(f'{label} — EXCLUDED (already in HubSpot, ID: {hubspot_id})')
                write_json(EXCLUDED_FILE, excluded)
                sleep_ms(100)
                continue
        except Exception as err:
            if str(err) == 'RATE_LIMITED':
                print(f'\nHubSpot RATE LIMITED. Halting. Re-run to resume.')
                print(f'Progress: {headline_excluded} headline + {hubspot_excluded} hubspot excluded, {passed} passed')
                sys.exit(1)
            print(f'{label} — HubSpot lookup error: {err} (keeping contact)')

        # Passed both filters
        filtered.append(contact)
        passed += 1
        print(f'{label} — PASSED')

        write_json(FILTERED_FILE, filtered)
        sleep_ms(100)

    print(f'\nDone.')
    print(f'  Total:                {len(contacts)}')
    print(f'  Excluded (headline):  {headline_excluded}')
    print(f'  Excluded (HubSpot):   {hubspot_excluded}')
    print(f'  Ready for enrichment: {passed}')
    print(f'  Total filtered:       {len(filtered)}')
    print(f'  Total excluded:       {len(excluded)}')


if __name__ == '__main__':
    main()
